### PR TITLE
Add deployment templates.

### PR DESCRIPTION
**Pull request by tgxworld**

### CODE REVIEW

File: deploy/templates/postgres.template.yml [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+  db_shared_buffers: "256MB"
+  db_work_mem: "10MB"
+  db_default_text_search_config: "pg_catalog.english"
+  db_name: rubybench
+  db_user: rubybench
+
+hooks:
+  before_code:
+    - replace:
+       filename: /etc/service/puma/run
```
Comment (tgxworld) [inline deploy/templates/postgres.template.yml]: This file will be create in `web.template.yml` with `# postgres`

File: deploy/templates/redis.template.yml [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+      hook: redis
+
+hooks:
+  before_code:
+    - exec:
+        background: true
+        cmd: exec chpst -u redis -U redis /usr/bin/redis-server /etc/redis/redis.conf
+  after_code:
+    - replace:
+       filename: /etc/service/puma/run
```
Comment (tgxworld) [inline deploy/templates/redis.template.yml]: This file will be create in `web.template.yml` with `# redis`

**Comment by tgxworld** [comment]
@SamSaffron I've gotten to used to `pups` now. Will be adding the `web.template.yml` over this wekk.

File: deploy/base/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    npm install uglify-js -g
+
+RUN add-apt-repository ppa:ubuntu-toolchain-r/test &&\
+    apt-get update &&\
+    apt-get install -y gcc-4.9 &&\
+    (cd /usr/bin && rm gcc && ln -s gcc-4.9 gcc) &&\
+    echo 'gem: --no-document' >> /usr/local/etc/gemrc &&\
+    mkdir /src && cd /src && git clone https://github.com/sstephenson/ruby-build.git &&\
+    cd /src/ruby-build && ./install.sh &&\
+    cd / && rm -rf /src/ruby-build && ruby-build 2.2.1 /usr/local &&\
```
Comment (tgxworld) [inline deploy/base/Dockerfile]: Using 2.2.1 here.

File: deploy/base/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+    add-apt-repository -y ppa:nginx/development &&\
+    apt-get install -y curl && curl http://apt.postgresql.org/pub/repos/apt/ACCC4CF8.asc | apt-key add - &&\
+    echo "deb http://apt.postgresql.org/pub/repos/apt/ trusty-pgdg main" | \
+        tee /etc/apt/sources.list.d/postgres.list &&\
+    curl -sL https://deb.nodesource.com/setup | sudo bash - &&\
+    apt-get -y update &&\
+    apt-get -y install build-essential git curl wget \
+                       libxslt-dev libcurl4-openssl-dev \
+                       libssl-dev libyaml-dev libtool \
+                       libxml2-dev libffi-dev gawk \
```
Comment (tgxworld) [inline deploy/base/Dockerfile]: Install `libffi-dev` otherwise installing Ruby 2.2.1 fails.

File: deploy/image/ruby-bench-web/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,15 @@
+FROM rubybench/rubybench_docker_base
+MAINTAINER Alan Guo Xiang Tan "https://twitter.com/tgx_world"
+
+RUN sudo apt-get install -y qt5-default libqt5webkit5-dev
```
Comment (tgxworld) [inline deploy/image/ruby-bench-web/Dockerfile]: For `capybara-webkit` to install successfully.

**Comment by tgxworld** [comment]
@SamSaffron Could you have a look, I'm not sure if I'm doing the right thing here. For `web.template.yml`, should I be following https://github.com/discourse/discourse_docker/blob/master/templates/web.template.yml or https://github.com/discourse/logster/blob/master/website/docker_container/logster.yml#L1?

File: deploy/templates/web.template.yml [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+    - exec: rm /etc/nginx/sites-enabled/default
+
+    - replace:
+        filename: /etc/nginx/nginx.conf
+        from: pid /run/nginx.pid;
+        to: daemon off;
+
+    - file:
+        path: /etc/nginx/conf.d/logster.conf
```
Comment (SamSaffron) [inline deploy/templates/web.template.yml]: should be rubybench.conf , probably should be checked in direct to rubybench web and just copied over to etc/nginx/conf.d instead ...

File: deploy/containers/app.yml [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+  db_name: rubybench_production
+  db_user: rubybench
+  home: /var/www/ruby-bench-web
+
+templates:
+  - "templates/redis.template.yml"
+  - "templates/postgres.template.yml"
+
+expose:
+  - "3000:3000"
```
Comment (tgxworld) [inline deploy/containers/app.yml]: @SamSaffron I think I got it to work because this is exposing the port directly. It doesn't seem like it is going through `nginx`. Not sure if I'm supposed to be doing this.

File: deploy/containers/app.yml [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+
+hooks:
+  after_postgres:
```
Comment (tgxworld) [inline deploy/containers/app.yml]: changed to run after_postgres so that I can migrate the database.

File: deploy/containers/app.yml [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+        from: |
+          # production:
+          #   <<: *default
+          #   database: ruby-bench-web_production
+          #   username: ruby-bench-web
+        to: |
+          production:
+            <<: *default
+            database: $db_name
+            username: $db_user
```
Comment (tgxworld) [inline deploy/containers/app.yml]: This is just a hack for now. I should probably store this in a config.

File: deploy/image/ruby-bench-web/Dockerfile [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+MAINTAINER Alan Guo Xiang Tan "https://twitter.com/tgx_world"
+
+RUN sudo apt-get update && sudo apt-get install -y qt5-default libqt5webkit5-dev
+
+RUN useradd rubybench -s /bin/bash -m -U &&\
+    mkdir -p /var/www && cd /var/www &&\
+      git clone https://github.com/ruby-bench/ruby-bench-web.git &&\
+      cd ruby-bench-web &&\
+        git config --global user.name "rubybench" &&\
+        git config --global user.email "https://twitter.com/tgx_world" &&\
```
Comment (tgxworld) [inline deploy/image/ruby-bench-web/Dockerfile]: I ran into problems where it asked me for my `git config` when doing `git pull`.

**Comment by tgxworld** [comment]
@SamSaffron I tested it on the server and managed to get it to work. However, I think I did something wrong with `nginx`. I removed the config entirely and it still works. Any idea what I might have done wrong? Thanks in advance :smile:

File: deploy/containers/app.yml (hunk 7 above)
Comment (SamSaffron) [inline deploy/containers/app.yml]: don't expose 3000, that is internal, expose port 80 NGINX will talk to unicorn on 3000

**Comment by SamSaffron** [comment]
this is looking so much better and simpler

**Comment by tgxworld** [comment]
@SamSaffron Ok managed to get Unicorn working with Nginx.

**Comment by SamSaffron** [comment]
this looks good to me, when can we make the move :) 

Really want to get rack-mini-profiler going on the web so we can confirm perf is outstanding :)

**Comment by tgxworld** [comment]
@SamSaffron  Added logging and sidekiq. :smile

File: deploy/containers/app.yml [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+  - "templates/postgres.template.yml"
+
+expose:
+  - "80:80"
+
+volumes:
+  - volume:
+      host: /var/discourse/discourse_docker/shared/rubybench
+      guest: /shared
+  - volume:
```
Comment (SamSaffron) [inline deploy/containers/app.yml]: you already mounted shared, no need for these two extra mounts

File: deploy/containers/app.yml [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+          #!/bin/sh
+          exec 2>&1
+          exec /usr/sbin/nginx
+
+    - file:
+       path: /etc/runit/3.d/01-nginx
+       chmod: "+x"
+       contents: |
+         #!/bin/bash
+         sv stop nginx
```
Comment (SamSaffron) [inline deploy/containers/app.yml]: why is this needed? runsvdir will already take care of this for you and it is stopped first, in stage 3 you just need to clean up mess that is left over like possibly pid files and such.

**Comment by SamSaffron** [comment]
merging in work in progress
